### PR TITLE
Submit IO in batches to reduce overhead

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We should submit multiple IO ops at once to reduce the syscall overhead.

Modifications:

- Submit multiple IO ops in batches
- Adjust default ringsize

Result:

Much better performance

### CODE REVIEW

**Comment by normanmaurer** [comment]
With this commit...

NIO echoserver:

```
norman@ubuntu-dev:~/Apps/rust_echo_bench$ ./run.sh 9090
Linux ubuntu-dev 5.8.5-iouringfix+ #2 SMP Thu Sep 3 12:58:01 CEST 2020 x86_64 x86_64 x86_64 GNU/Linux
pid 126331's current affinity list: 0
pid 126331's new affinity list: 0
    Finished release [optimized] target(s) in 0.01s
     Running `target/release/echo_bench --address 'localhost:9090' --number 500 --duration 30 --length 128`
Benchmarking: localhost:9090
500 clients, running 128 bytes, 30 sec.

Speed: 78744 request/sec, 78744 response/sec
Requests: 2362321
Responses: 2362320
```

IO_URING based echo server:

```
./run.sh 8080
Linux ubuntu-dev 5.8.5-iouringfix+ #2 SMP Thu Sep 3 12:58:01 CEST 2020 x86_64 x86_64 x86_64 GNU/Linux
pid 138209's current affinity list: 0
pid 138209's new affinity list: 0
    Finished release [optimized] target(s) in 0.00s
     Running `target/release/echo_bench --address 'localhost:8080' --number 500 --duration 30 --length 128`
Benchmarking: localhost:8080
500 clients, running 128 bytes, 30 sec.

Speed: 124256 request/sec, 124256 response/sec
Requests: 3727709
Responses: 3727708
```

**Comment by normanmaurer** [comment]
/cc @njhill @franz1981 @chrisvest 

After this one is merged I have another one change pending that @chrisvest suggested that improve the perf again :)

**Comment by njhill** [comment]
@normanmaurer I expect you could get another big boost from some additional small changes to the event loop `run()` method like I did in the [other PR](https://github.com/1Jo1/netty/pull/25/files#diff-edc47287cb4c12155b5456eb05260bcbR128).

**Comment by normanmaurer** [comment]
@njhill let's get this one merged and then you can do another PR on top of this changeset ? =

**Review by njhill — APPROVED** [review]

**Review by franz1981 — APPROVED** [review]

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketGatheringWriteTest.java [hunk 1]
```diff
@@ -99,6 +100,7 @@ public void testGatheringWriteWithComposite(ServerBootstrap sb, Bootstrap cb) th
 
     // Test for https://github.com/netty/netty/issues/2647
     @Test
+    @Ignore
```
Comment (chrisvest) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketGatheringWriteTest.java]: Is this a debugging left-over, or are we ignoring this test for now?

File: transport-native-io_uring/src/test/java/io/netty/channel/uring/IOUringSocketHalfClosedTest.java [hunk 2]
```diff
@@ -27,4 +31,12 @@
     protected List<TestsuitePermutation.BootstrapComboFactory<ServerBootstrap, Bootstrap>> newFactories() {
         return IOUringSocketTestPermutation.INSTANCE.socket();
     }
+
+    @Ignore
+    @Test
+    public void testAutoCloseFalseDoesShutdownOutput() throws Throwable {
```
Comment (chrisvest) [inline transport-native-io_uring/src/test/java/io/netty/channel/uring/IOUringSocketHalfClosedTest.java:37]: And this one?

**Review by chrisvest — COMMENTED** [review]
Had some questions about the ignored tests

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketGatheringWriteTest.java (hunk 1 above)
Comment (normanmaurer) [reply testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketGatheringWriteTest.java]: yep ... sorry let me remove this one.

File: transport-native-io_uring/src/test/java/io/netty/channel/uring/IOUringSocketHalfClosedTest.java (hunk 2 above)
Comment (normanmaurer) [reply transport-native-io_uring/src/test/java/io/netty/channel/uring/IOUringSocketHalfClosedTest.java:37]: I ignored this one for now as it makes some assumption about execution that is not really valid and more an implementation detail.

**Comment by njhill** [comment]
@normanmaurer here it is: https://github.com/1Jo1/netty/pull/33 ... I will rebase once this one is merged

**Review by chrisvest — APPROVED** [review]

**Pull request merged**